src/validation.h

```diff
@@ -461,27 +461,89 @@ struct CBlockIndexWorkComparator
 };
 
 /**
- * CChainState stores and provides an API to update our local knowledge of the
- * current best chain and header tree.
+ * Maintains a tree of blocks (stored in `m_block_index`) which is consulted
+ * to determine where the most-work tip is.
  *
- * It generally provides access to the current block tree, as well as functions
- * to provide new data, which it will appropriately validate and incorporate in
- * its state as necessary.
+ * This data is used mostly in `CChainState` - information about, e.g.,
+ * candidate tips is not maintained here.
+ */
+class BlockManager {
+public:
+    BlockMap m_block_index GUARDED_BY(cs_main);
+
+    /** In order to efficiently track invalidity of headers, we keep the set of
+      * blocks which we tried to connect and found to be invalid here (ie which
+      * were set to BLOCK_FAILED_VALID since the last restart). We can then
+      * walk this set and check if a new header is a descendant of something in
+      * this set, preventing us from having to walk m_block_index when we try
+      * to connect a bad block and fail.
+      *
+      * While this is more complicated than marking everything which descends
+      * from an invalid block as invalid at the time we discover it to be
+      * invalid, doing so would require walking all of m_block_index to find all
+      * descendants. Since this case should be very rare, keeping track of all
+      * BLOCK_FAILED_VALID blocks in a set should be just fine and work just as
+      * well.
+      *
+      * Because we already walk m_block_index in height-order at startup, we go
+      * ahead and mark descendants of invalid blocks as FAILED_CHILD at that time,
+      * instead of putting things in this set.
+      */
+    std::set<CBlockIndex*> m_failed_blocks;
+
+public:
```

Comment:
In commit e2c878bea3 refactoring: move block metadata structures into BlockManager

style-nit: Duplicate `public`

Also, wouldn't it be clearer to not expose the block index via CChainState::BlockIndex? For me this is confusing because there is one blockindex, but several chainstates that could use it.

I suggest the following diff, which adds a `BlockManager::BlockIndex` method (diff is on top of the current pull request, not the commit where I left this comment):

```diff
diff --git a/src/init.cpp b/src/init.cpp
index 5097a81f08..2235db5ba1 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -1563,7 +1563,7 @@ bool AppInitMain(InitInterfaces& interfaces)
 
                 // If the loaded chain has a wrong genesis, bail out immediately
                 // (we're likely using a testnet datadir, or the other way around).
-                if (!::ChainstateActive().BlockIndex().empty() &&
+                if (!g_chainman.BlockIndex().empty() &&
                         !LookupBlockIndex(chainparams.GetConsensus().hashGenesisBlock)) {
                     return InitError(_("Incorrect or no genesis block found. Wrong datadir for network?"));
                 }
@@ -1798,7 +1798,7 @@ bool AppInitMain(InitInterfaces& interfaces)
     //// debug print
     {
         LOCK(cs_main);
-        LogPrintf("m_block_index.size() = %u\n", ::ChainstateActive().BlockIndex().size());
+        LogPrintf("m_block_index.size() = %u\n", g_chainman.BlockIndex().size());
         chain_active_height = ::ChainActive().Height();
     }
     LogPrintf("nBestHeight = %d\n", chain_active_height);
diff --git a/src/rpc/blockchain.cpp b/src/rpc/blockchain.cpp
index 6cdd11233a..d8e40a5d56 100644
--- a/src/rpc/blockchain.cpp
+++ b/src/rpc/blockchain.cpp
@@ -1378,7 +1378,7 @@ static UniValue getchaintips(const JSONRPCRequest& request)
     std::set<const CBlockIndex*> setOrphans;
     std::set<const CBlockIndex*> setPrevs;
 
-    for (const std::pair<const uint256, CBlockIndex*>& item : ::ChainstateActive().BlockIndex())
+    for (const std::pair<const uint256, CBlockIndex*>& item : g_chainman.BlockIndex())
     {
         if (!::ChainActive().Contains(item.second)) {
             setOrphans.insert(item.second);
diff --git a/src/validation.cpp b/src/validation.cpp
index f7884759ad..ebdd3ba1bc 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -1044,11 +1044,6 @@ bool CChainState::IsInitialBlockDownload() const
     return false;
 }
 
-BlockMap& CChainState::BlockIndex()
-{
-    return m_blockman.m_block_index;
-}
-
 CBlockIndex *pindexBestForkTip = nullptr, *pindexBestForkBase = nullptr;
 
 static void AlertNotify(const std::string& strMessage)
@@ -1695,8 +1690,8 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
         //  relative to a piece of software is an objective fact these defaults can be easily reviewed.
         // This setting doesn't force the selection of any particular chain but makes validating some faster by
         //  effectively caching the result of part of the verification.
-        BlockMap::const_iterator  it = BlockIndex().find(hashAssumeValid);
-        if (it != BlockIndex().end()) {
+        BlockMap::const_iterator  it = m_blockman.BlockIndex().find(hashAssumeValid);
+        if (it != m_blockman.BlockIndex().end()) {
             if (it->second->GetAncestor(pindex->nHeight) == pindex &&
                 pindexBestHeader->GetAncestor(pindex->nHeight) == pindex &&
                 pindexBestHeader->nChainWork >= nMinimumChainWork) {
@@ -2787,8 +2782,8 @@ bool CChainState::InvalidateBlock(CValidationState& state, const CChainParams& c
 
         // The resulting new best tip may not be in setBlockIndexCandidates anymore, so
         // add it again.
-        BlockMap::iterator it = BlockIndex().begin();
-        while (it != BlockIndex().end()) {
+        BlockMap::iterator it = m_blockman.BlockIndex().begin();
+        while (it != m_blockman.BlockIndex().end()) {
             if (it->second->IsValid(BLOCK_VALID_TRANSACTIONS) && it->second->HaveTxsDownloaded() && !setBlockIndexCandidates.value_comp()(it->second, m_chain.Tip())) {
                 setBlockIndexCandidates.insert(it->second);
             }
@@ -2815,8 +2810,8 @@ void CChainState::ResetBlockFailureFlags(CBlockIndex *pindex) {
     int nHeight = pindex->nHeight;
 
     // Remove the invalidity flag from this block and all its descendants.
-    BlockMap::iterator it = BlockIndex().begin();
-    while (it != BlockIndex().end()) {
+    BlockMap::iterator it = m_blockman.BlockIndex().begin();
+    while (it != m_blockman.BlockIndex().end()) {
         if (!it->second->IsValid() && it->second->GetAncestor(nHeight) == pindex) {
             it->second->nStatus &= ~BLOCK_FAILED_MASK;
             setDirtyBlockIndex.insert(it->second);
@@ -3956,7 +3951,7 @@ bool CChainState::LoadChainTip(const CChainParams& chainparams)
 
     if (tip && tip->GetBlockHash() == coins_cache.GetBestBlock()) return true;
 
-    if (coins_cache.GetBestBlock().IsNull() && BlockIndex().size() == 1) {
+    if (coins_cache.GetBestBlock().IsNull() && m_blockman.BlockIndex().size() == 1) {
         // In case we just added the genesis block, connect it now, so
         // that we always have a tip when we return.
         LogPrintf("%s: Connecting genesis block...\n", __func__);
@@ -4139,16 +4134,16 @@ bool CChainState::ReplayBlocks(const CChainParams& params)
     const CBlockIndex* pindexNew;            // New tip during the interrupted flush.
     const CBlockIndex* pindexFork = nullptr; // Latest block common to both the old and the new tip.
 
-    if (BlockIndex().count(hashHeads[0]) == 0) {
+    if (m_blockman.BlockIndex().count(hashHeads[0]) == 0) {
         return error("ReplayBlocks(): reorganization to unknown block requested");
     }
-    pindexNew = BlockIndex()[hashHeads[0]];
+    pindexNew = m_blockman.BlockIndex()[hashHeads[0]];
 
     if (!hashHeads[1].IsNull()) { // The old tip is allowed to be 0, indicating it's the first flush.
-        if (BlockIndex().count(hashHeads[1]) == 0) {
+        if (m_blockman.BlockIndex().count(hashHeads[1]) == 0) {
             return error("ReplayBlocks(): reorganization from unknown block requested");
         }
-        pindexOld = BlockIndex()[hashHeads[1]];
+        pindexOld = m_blockman.BlockIndex()[hashHeads[1]];
         pindexFork = LastCommonAncestor(pindexOld, pindexNew);
         assert(pindexFork != nullptr);
     }
@@ -4235,7 +4230,7 @@ bool CChainState::RewindBlockIndex(const CChainParams& params)
     // blocks will be dealt with below (releasing cs_main in between).
     {
         LOCK(cs_main);
-        for (const auto& entry : BlockIndex()) {
+        for (const auto& entry : m_blockman.BlockIndex()) {
             if (IsWitnessEnabled(entry.second->pprev, params.GetConsensus()) && !(entry.second->nStatus & BLOCK_OPT_WITNESS) && !m_chain.Contains(entry.second)) {
                 EraseBlockData(entry.second);
             }
@@ -4373,7 +4368,7 @@ bool LoadBlockIndex(const CChainParams& chainparams) EXCLUSIVE_LOCKS_REQUIRED(cs
     if (!fReindex) {
         bool ret = LoadBlockIndexDB(chainparams);
         if (!ret) return false;
-        needs_init = g_chainman.m_blockman.m_block_index.empty();
+        needs_init = g_chainman.m_blockman.BlockIndex().empty();
     }
 
     if (needs_init) {
@@ -4553,17 +4548,17 @@ void CChainState::CheckBlockIndex(const Consensus::Params& consensusParams)
     // so we have the genesis block in BlockIndex() but no active chain. (A few of the tests when
     // iterating the block tree require that m_chain has been initialized.)
     if (m_chain.Height() < 0) {
-        assert(BlockIndex().size() <= 1);
+        assert(m_blockman.BlockIndex().size() <= 1);
         return;
     }
 
     // Build forward-pointing map of the entire block tree.
     std::multimap<CBlockIndex*,CBlockIndex*> forward;
-    for (const std::pair<const uint256, CBlockIndex*>& entry : BlockIndex()) {
+    for (const std::pair<const uint256, CBlockIndex*>& entry : m_blockman.BlockIndex()) {
         forward.insert(std::make_pair(entry.second->pprev, entry.second));
     }
 
-    assert(forward.size() == BlockIndex().size());
+    assert(forward.size() == m_blockman.BlockIndex().size());
 
     std::pair<std::multimap<CBlockIndex*,CBlockIndex*>::iterator,std::multimap<CBlockIndex*,CBlockIndex*>::iterator> rangeGenesis = forward.equal_range(nullptr);
     CBlockIndex *pindex = rangeGenesis.first->second;
diff --git a/src/validation.h b/src/validation.h
index 84605e957c..20a90cda0b 100644
--- a/src/validation.h
+++ b/src/validation.h
@@ -465,8 +465,12 @@ struct CBlockIndexWorkComparator
  * candidate tips is not maintained here.
  */
 class BlockManager {
-public:
     BlockMap m_block_index GUARDED_BY(cs_main);
+public:
+    BlockMap& BlockIndex() EXCLUSIVE_LOCKS_REQUIRED(::cs_main)
+    {
+        return m_block_index;
+    }
 
     /** In order to efficiently track invalidity of headers, we keep the set of
       * blocks which we tried to connect and found to be invalid here (ie which
@@ -488,7 +492,6 @@ public:
       */
     std::set<CBlockIndex*> m_failed_blocks;
 
-public:
     /*
      * All pairs A->B, where A (or one of its ancestors) misses transactions, but B has transactions.
      * Pruned nodes may have entries where B is missing data.
@@ -775,9 +778,6 @@ public:
     /** Check whether we are doing an initial block download (synchronizing from disk or network) */
     bool IsInitialBlockDownload() const;
 
-    //! @returns the map of blocks that this chainstate is aware of.
-    BlockMap& BlockIndex();
-
     /** Update the chain tip based on database information. */
     bool LoadChainTip(const CChainParams& chainparams) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
 
@@ -1040,7 +1040,7 @@ public:
 
     BlockMap& BlockIndex() EXCLUSIVE_LOCKS_REQUIRED(::cs_main)
     {
-        return m_blockman.m_block_index;
+        return m_blockman.BlockIndex();
     }
 
     bool IsSnapshotActive() const
diff --git a/src/wallet/test/wallet_tests.cpp b/src/wallet/test/wallet_tests.cpp
index 91618525fd..afb619f13f 100644
--- a/src/wallet/test/wallet_tests.cpp
+++ b/src/wallet/test/wallet_tests.cpp
@@ -272,7 +272,7 @@ static int64_t AddTx(CWallet& wallet, uint32_t lockTime, int64_t mockTime, int64
     if (blockTime > 0) {
         LockAnnotation lock(::cs_main);
         auto locked_chain = wallet.chain().lock();
-        auto inserted = ::ChainstateActive().BlockIndex().emplace(GetRandHash(), new CBlockIndex);
+        auto inserted = g_chainman.BlockIndex().emplace(GetRandHash(), new CBlockIndex);
         assert(inserted.second);
         const uint256& hash = inserted.first->first;
         block = inserted.first->second;

Reply:
Yeah I like your suggestion, thanks! Will work it into the commits.